src/tribler/core/components/ipv8/eva/protocol.py

```diff
@@ -185,46 +235,74 @@ async def on_write_request_packet(self, peer: Peer, payload: WriteRequest):
-            self._finish_with_error(transfer, e)
+        transfer = IncomingTransfer(
+            container=self.incoming,
+            peer=peer,
+            info=payload.info,
+            nonce=payload.nonce,
+            data_size=payload.data_size,
+            settings=self.settings,
+            parent_task_group=self.task_group,
+            send_message=self.send_message,
+            on_complete=self.on_receive,
+            on_error=self.on_error,
+        )
+
+        if await self.check_transfer_correctness(transfer):
+            transfer.start()
+
+    @message_handler(ReadRequest)
+    async def on_read_request(self, peer: Peer, payload: ReadRequest):
+        logger.debug(f'On read request. Peer: {peer}. Info: {payload.info}.')
+        if peer in self.outgoing:
             return
 
-        if self._is_simultaneously_served_transfers_limit_exceeded():
-            exception = TransferLimitException('Maximum simultaneous transfers limit exceeded')
-            self._finish_with_error(transfer, exception)
+        data = b''
+        exception = None
+        if self.on_request:
+            try:
+                data = await self.on_request(peer, payload.info)
+            except Exception as e:  # pylint: disable=broad-except
+                exception = e
+        else:
+            exception = NotImplementedError('on_request callback is not defined')
+
+        transfer = OutgoingTransfer(
+            container=self.outgoing,
+            peer=peer,
+            info=payload.info,
+            data=data,
+            data_size=len(data),
+            nonce=payload.nonce,
+            settings=self.settings,
+            parent_task_group=self.task_group,
+            send_message=self.send_message,
+            on_complete=self.on_send_complete,
+            on_error=self.on_error
+        )
+
+        if exception:
+            message = f'{exception.__class__.__name__}: {exception}'
+            logger.warning(f'Exception during data retrieval. {message}')
+            await self._finish_with_error(transfer, RequestRejected(message, transfer))
             return
 
-        self.incoming[peer] = transfer
-
-        self.community.register_anonymous_task('eva_terminate_by_timeout', transfer.terminate_by_timeout_task)
-        self.community.register_anonymous_task('eva_resend_acknowledge', self._resend_acknowledge_task, transfer)
-        self.send_message(peer, transfer.make_acknowledgement())
+        if await self.check_transfer_correctness(transfer):
+            transfer.start()
 
     @message_handler(Acknowledgement)
     async def on_acknowledgement_packet(self, peer: Peer, payload: Acknowledgement):
         logger.debug(f'On acknowledgement({payload.number}). Window size: {payload.window_size}. Peer: {peer}.')
 
-        transfer = self.outgoing.get(peer)
+        transfer = self._get_transfer(peer=peer, incoming=False, nonce=payload.nonce)
```

Comment:
**nitpick:** it is better to have two separate methods, `_get_incoming_transfer` and `_get_outgoing_transfer`, to have a correct transfer type (`IncomingTransfer`/`OutgoingTransfer` instead of the base `Transfer` class)

Reply:
It is an option, yes.

But this method is also used here (so, from my POV it is better to leave it as is):

```python
    @message_handler(Error)
    async def on_error_packet(self, peer: Peer, error: Error):
        message = error.message.decode('utf-8')
        logger.debug(f'On error. Peer: {peer}. Message: "{message}"')

        transfer = self._get_transfer(peer=peer, incoming=not error.incoming, nonce=error.nonce)
        if transfer:
            exception_cls = to_class(error.code)
            transfer.finish(exception=exception_cls(message, transfer, remote=True))
            self.scheduler.send_scheduled()
```